### ISSUE TITLE
Batched force/energy evaluation is not bitwise reproducible (batch > 2, CUDA)

### ISSUE THREAD

**Issue by isayev** [isayevlab/aimnetcentral#93]
## Summary

Repeated **identical** calls to `AIMNet2Calculator.eval(..., forces=True)` return different energies and forces once the batch dimension exceeds 2. `B=1` and `B=2` are bitwise reproducible; `B>=4` is not.

The magnitudes are tiny (energy ~2.4e-07, forces ~1.2e-10), but they are amplified by iterative optimizers into macroscopic differences — see "Downstream impact" below.

## Reproducer

```python
import torch
from aimnet.calculators import AIMNet2Calculator

DEV = "cuda"
NUMBERS = [6, 7, 6, 7, 6, 6, 6, 8, 7, 6, 8, 7, 6, 6, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
COORDS = [
    [1.1610, 0.3860, -0.0066], [0.2500, 1.3900, -0.0021], [-1.0940, 1.1370, 0.0028],
    [-1.5230, -0.1220, 0.0025], [-0.6350, -1.1450, -0.0018], [0.7220, -0.9210, -0.0065],
    [1.6350, -2.0270, -0.0113], [2.8460, -1.9210, -0.0155], [1.1250, -3.3080, -0.0109],
    [1.9250, -4.4550, -0.0157], [3.1370, -4.3610, -0.0202], [-0.2860, -3.5150, -0.0062],
    [-0.7960, -4.8800, -0.0057], [-2.7010, 2.1440, 0.0079], [0.6020, 2.4060, -0.0043],
    [-1.8090, 1.9330, 0.0072], [2.2410, 0.4300, -0.0100], [1.4310, -5.4180, -0.0154],
    [-1.8870, -4.9160, -0.0021], [-0.4210, -5.3980, 0.8760], [-0.4150, -5.4040, -0.8830],
    [-3.1710, 2.0180, 0.9860], [-3.1750, 2.0130, -0.9670], [-2.3320, 3.1690, 0.0130],
]

calc = AIMNet2Calculator("aimnet2")
z = torch.tensor(NUMBERS, device=DEV)
c = torch.tensor(COORDS, dtype=torch.float32, device=DEV)

def ef(B):
    out = calc.eval({
        "coord": c.unsqueeze(0).expand(B, -1, -1).contiguous(),
        "numbers": z.unsqueeze(0).expand(B, -1).contiguous(),
        "charge": torch.zeros(B, device=DEV),
    }, forces=True)
    return out["energy"], out["forces"]

for B in (1, 2, 4, 8, 16, 32):
    ef(B)                      # warm this shape
    e0, f0 = ef(B)             # reference
    de = df = 0.0
    for _ in range(8):
        e, f = ef(B)
        de = max(de, float((e - e0).abs().max()))
        df = max(df, float((f - f0).abs().max()))
    print(f"B={B:>3}  max|dE|={de:.3e}  max|dF|={df:.3e}  "
          f"{'reproducible' if df == 0 else 'NOT reproducible'}")
```

### Observed

```
B=  1  max|dE|=0.000e+00  max|dF|=0.000e+00  reproducible
B=  2  max|dE|=0.000e+00  max|dF|=0.000e+00  reproducible
B=  4  max|dE|=2.384e-07  max|dF|=1.164e-10  NOT reproducible
B=  8  max|dE|=2.384e-07  max|dF|=1.164e-10  NOT reproducible
B= 16  max|dE|=2.384e-07  max|dF|=1.164e-10  NOT reproducible
B= 32  max|dE|=2.384e-07  max|dF|=1.164e-10  NOT reproducible
```

Expected: identical inputs produce identical outputs.

## What I ruled out

- **Not a PyTorch nondeterministic op.** `torch.use_deterministic_algorithms(True)` (error mode, with `CUBLAS_WORKSPACE_CONFIG=:4096:8`) raises **no error** anywhere on this path.
- **Not the Warp `conv_sv_2d_sp` kernel.** `aimnet/kernels/conv_sv_2d_sp_wp.py:89` uses `wp.atomic_add` on the gradient, which looked like the obvious candidate. I instrumented the dispatch site in `aimnet/modules/aev.py:162` and it is invoked **0 times** for this model at B=1/2/8/32, in both float32 and float64 — the `d2features` branch is not taken.
- **Not warmup / first-call algorithm selection.** The reproducer warms each shape before taking its reference.
- **Not `torch.compile`.** Runs with `TORCH_COMPILE_DISABLE=1` and `compile_model=False`.

## Where it might be

Profiling the batched force path (`torch.profiler`) shows a custom op with a generated backward:

```
GeneratedBackwardFor_nvalchemiops__naive_batch_neighbor_matrix_no_pbc_default
```

The name lines up with the observed split — a *batch* neighbor-matrix kernel, where `B=1` takes a different route — and an external CUDA op would also explain why `use_deterministic_algorithms` cannot see it. **I have not confirmed this**; it would need `nvalchemiops` sources to pin the atomic. Treat it as a lead, not a diagnosis.

## Downstream impact

This is not merely cosmetic at 1e-10. Feeding these forces to an iterative optimizer amplifies them:

- Single-molecule (`B=1`) geometry optimization is bitwise reproducible across repeated runs **and across processes**.
- Batched optimization is not.
- In a reaction campaign (batch size 32), two byte-identical invocations diverged on **7 of 66** rows, with transition-state energies differing by up to **2.05 eV**. One row converged in 74 steps at `fmax=0.06` in one run and wandered 1000 steps to `fmax=4.4` in the other, from identical input.

The practical consequence is that batched A/B comparisons need a same-config control arm to be interpretable, since run-to-run noise can exceed the effect being measured.

## Environment

- `aimnet` 0.2.0.post1.dev34+g8092749f9
- `torch` 2.9.1+cu128
- NVIDIA L40S, CUDA 12.9, driver 13.2
- Warp 1.12.0
- `nvalchemiops` installed

## Question

Is batched bitwise reproducibility intended to hold? If the atomics are a deliberate performance trade, an opt-in deterministic path (or a documented note that batched forces are not reproducible) would help downstream consumers that need reproducible optimization trajectories.

**Comment by isayev** [comment]
Diagnosed. The nondeterminism is not in aimnet's own code — the NN core is bitwise reproducible at every batch size (verified by toggling external modules: with `needs_dispersion=False` every B is stable). It enters through the external DFT-D3 path, in two independent layers inside nvalchemiops:

1. `neighbor_list(method="batch_naive")` returns per-row neighbor *sets* stably but in nondeterministic within-row *order* (atomic slot assignment), so downstream float32 pair sums iterate in varying order.
2. The `dftd3` kernel itself varies at accumulation-rounding level (up to 3.7e-09 Hartree over 10 calls) even with a **frozen** neighbor matrix — its own atomics race.

Both are performance trades in the CUDA kernels; filed upstream with minimal reproducers: NVIDIA/nvalchemi-toolkit-ops#135.

Two updates since the original report:

- Under nvalchemi-toolkit-ops 0.4.0 (bumped in #100), the B<=2 safe zone is gone: B=2 now also drifts (~1e-7 eV). Same noise floor, wider surface — consistent with upstream's rework of the direct-output force paths.
- Practical mitigation landed in #101: `AIMNet2Calculator(..., deterministic=True)` routes external D3 (and DSF Coulomb) through the differentiable pure-torch paths, giving bitwise-identical repeated evaluations on the same machine/build at no measurable cost for batched workloads (19.9 vs 19.5 ms/eval at B=32). That makes batched optimization trajectories repeatable, which was the real pain here.

Leaving this open to track the upstream kernel question; the flag is the workaround in the meantime.

**Comment by isayev** [comment]
Closing: the practical need here — repeatable batched optimization trajectories — is resolved by `AIMNet2Calculator(..., deterministic=True)` (#101), which routes external D3 and DSF Coulomb through the pure-torch paths. Re-verified on current main: the default path still drifts at B=8 (max|dE| = 2.4e-07 eV, 1–2 float32 ulps of a partial sum), and the flag is bitwise-identical across repeated evaluations at no measurable cost.

On the magnitude itself: 2.4e-07 eV is ~5e-06 kcal/mol, five-plus orders below the model's accuracy against DFT — as an accuracy matter this is immaterial, so there is no case for slowing the default kernels. The trajectory divergence in the original report is chaotic amplification through saddle-point searches, which any ulp-scale perturbation (different GPU, different torch build) would trigger equally; the fix for that is the deterministic flag, not deterministic kernels.

The remaining kernel-level question — stable within-row neighbor ordering via row sorting, which upstream agrees is the right approach — is tracked in NVIDIA/nvalchemi-toolkit-ops#135 and doesn't need this issue open.
